actors/miner/src/vesting_state.rs

```diff
@@ -1,54 +1,120 @@
 }
 
 impl VestingFunds {
     pub fn new() -> Self {
-        Default::default()
+        Self(None)
+    }
+
+    pub fn load(&self, store: &impl Blockstore) -> Result<Vec<VestingFund>, ActorError> {
+        let Some(this) = &self.0 else { return Ok(Vec::new()) };
+        let mut funds: Vec<_> = store
+            .get_cbor(&this.tail)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load the vesting funds")?
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "missing vesting funds state")?;
+
+        // NOTE: we allow head to be drawn down to zero through fees/penalties. However, when
+        // inspecting the vesting table, we never want to see a "zero" entry so we skip it in
+        // that case.
+        if this.head.amount.is_positive() {
+            funds.insert(0, this.head.clone());
+        }
+        Ok(funds)
+    }
+
+    fn save(
+        &mut self,
+        store: &impl Blockstore,
+        funds: impl IntoIterator<Item = VestingFund>,
+    ) -> Result<(), ActorError> {
+        let mut funds = funds.into_iter();
+        let Some(head) = funds.next() else {
+            self.0 = None;
+            return Ok(());
+        };
+
+        let tail = store
+            .put_cbor(&funds.collect_vec(), Code::Blake2b256)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to store the vesting funds")?;
+
+        self.0 = Some(VestingFundsInner { head, tail });
+        Ok(())
     }
 
-    pub fn unlock_vested_funds(&mut self, current_epoch: ChainEpoch) -> TokenAmount {
-        // TODO: the funds are sorted by epoch, so we could do a binary search here
-        let i = self
-            .funds
-            .iter()
-            .position(|fund| fund.epoch >= current_epoch)
-            .unwrap_or(self.funds.len());
+    fn can_vest(&self, current_epoch: ChainEpoch) -> bool {
+        self.0.as_ref().map(|v| v.head.epoch < current_epoch).unwrap_or(false)
+    }
 
-        self.funds.drain(..i).map(|f| f.amount).sum()
+    pub fn unlock_vested_funds(
+        &mut self,
+        store: &impl Blockstore,
+        current_epoch: ChainEpoch,
+    ) -> Result<TokenAmount, ActorError> {
+        if !self.can_vest(current_epoch) {
+            return Ok(TokenAmount::zero());
+        }
+        let mut funds = self.load(store)?.into_iter().peekable();
+        let unlocked = take_vested(&mut funds, current_epoch);
+        self.save(store, funds)?;
+        Ok(unlocked)
     }
 
+    // Adds locked funds and unlocks everything that has already vested.
     pub fn add_locked_funds(
         &mut self,
+        store: &impl Blockstore,
         current_epoch: ChainEpoch,
         vesting_sum: &TokenAmount,
         proving_period_start: ChainEpoch,
         spec: &VestSpec,
-    ) {
+    ) -> Result<TokenAmount, ActorError> {
         // Quantization is aligned with when regular cron will be invoked, in the last epoch of deadlines.
```

Comment:
can you sanity check this comment @Stebalien, it doesn't seem true to me; we `offset` at `proving_period_start`

Reply:
It's worse:

1. We round up to the next proving period start (or 12 hours from then). So, deadline start not deadline end (as you noted).
2. Then we unlock the epoch _after_ that. I.e., we unlock everything where `vest_epoch < current_epoch`.

So, really, I _think_ the actual unlock will happen the deadline after.

(I should write a better test)

Reply:
Here's a diff for you for the top of this function:

```rust
        // Quantization is aligned with when the deadline start epoch, but as automatic unlocking
        // happens at the deadline end epoch, there will be a delay of one whole deadline.
        let quant = QuantSpec { unit: spec.quantization, offset: proving_period_start };
```

where `quant` can be used below in the iterator at `let vest_epoch = quant.quantize_up(epoch);`.

Reply:
done.

Reply:
Er, well, we still haven't resolved the rest of this discussion.

Reply:
^ there was a comment update suggestion in that too; if you get around to writing a test for it then you could come back and make the comment accurate

Reply:
Ok, I've documented the current behavior. I'd still like to consider changing this behavior but... it's not terrible.